Do not enforce new reserved value in PSBT midstep

Kixunil opened the pull request:
This change avoids enforcing new reserved value when PSBT funding is not
finished yet as new inputs and outputs may still be added that could
change the outcome of the check.

This originally failed in the scenario when funding a channel from
external wallet *and depositing to on-chain wallet* was done
simultaneously in a single transaction. If such transaction confirms
then reserved UTXO is guaranteed to be available but the check didn't
take it into account.

The enforcement still occurs in the final step of PSBT funding flow, so
it is safe. It also occurs in case of non-PSBT funding.

This partially fixes https://github.com/lightningnetwork/lnd/pull/5363#issuecomment-875766968
For full fix it'd be great to be able to optionally avoid anchor commitments (to improve privacy).

I will look at tests later, I mainly wanted to get initial feedback on this. If anyone would like to help with writing tests I'd be happy to accept. :)

#### Pull Request Checklist

- [x] If this is your first time contributing, we recommend you read the [Code
  Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md)
- [ ] All changes are Go version 1.12 compliant
- [x] The code being submitted is commented according to [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation)
- [ ] For new code: Code is accompanied by tests which exercise both
  the positive and negative (error paths) conditions (if applicable)
- [x] For bug fixes: Code is accompanied by new tests which trigger
  the bug being fixed to prevent regressions
- [ ] Any new logging statements use an appropriate subsystem and
  logging level
- [x] Code has been formatted with `go fmt`
- [ ] Protobuf files (`lnrpc/**/*.proto`) have been formatted with
  `make rpc-format` and compiled with `make rpc`
- [ ] New configuration flags have been added to `sample-lnd.conf`
- [x] For code and documentation: lines are wrapped at 80 characters
  (the tab character should be counted as 8 characters, not 4, as some IDEs do
  per default)
- [ ] Running `make check` does not fail any tests
- [ ] Running `go vet` does not report any issues
- [x] Running `make lint` does not report any **new** issues that did not
  already exist
- [ ] All commits build properly and pass tests. Only in exceptional
  cases it can be justifiable to violate this condition. In that case, the
  reason should be stated in the commit message.
- [x] Commits have a logical structure according to [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure)

Kixunil commented:
Ping, could someone check if this is at least going in the right direction?

guggero commented on lnwallet/wallet.go:
```diff
@@ -685,9 +685,16 @@ func (l *LightningWallet) handleFundingReserveRequest(req *InitFundingReserveMsg
 	// If no chanFunder was provided, then we'll assume the default
 	// assembler, which is backed by the wallet's internal coin selection.
 	if req.ChanFunder == nil {
+		enforceNewReservedValue = true
```
There are three types of `ChanFunder`: PSBT, canned (used for Pool) and the wallet assembler. We only want to skip this check for the wallet assembler. So I think we should use the default value `true` and only disable it for the PSBT funder (with a type check).

Kixunil replied:
I'm not sure I understood you correctly, so I added another commit. Is that what you had in mind?

guggero replied:
Yes, this is what I meant. But it should be done in the same commit.

guggero commented on lnwallet/wallet.go:
```diff
@@ -795,13 +802,15 @@ func (l *LightningWallet) handleFundingReserveRequest(req *InitFundingReserveMsg
-	err = l.enforceNewReservedValue(fundingIntent, isPublic, hasAnchors)
-	if err != nil {
-		fundingIntent.Cancel()
+	if enforceNewReservedValue {
```
So the comment above about PSBTs was incorrect before this change?

Kixunil replied:
Yes.

guggero reviewed (commented):
The PR must've slipped below the radar, sorry for the delay.
I think this is going into the right direction. Especially when reading that comment I think that was the intended behavior from the beginning but a refactor/fix somehow changed it.

This should be covered by an integration test I think, just to make sure nothing else in the funding flow breaks.

Kixunil commented:
Decided to add a note about only closing `quit` channel from `shutdown()` function.

Kixunil commented:
I agree it should have test. Unfortunately I have literally zero experience with writing tests in Go and LND. If you could give me a few bullet points on where to put it and how to structure it that'd be very helpful.

guggero commented:
See `lntest/itest/lnd_psbt_test.go`, there we already have a test that asserts the default PSBT channel funding flow. We could add an additional test that makes sure the reserves are checked in the correct step. Or extend the existing test with some additional checks.

Kixunil commented:
I looked into it and it seems `rpctest.Harness` doesn't have public methods for creating and broadcasting transactions separately. It seems it could be solved by adding `CreateTransaction` and `SendRawTransaction` that'd pass arguments to the internal wallet.

Should I file a PR against https://github.com/btcsuite/btcd/ first or is there some other process I should follow?

Kixunil commented:
Is there a way to run a specific test only? The tests are slow and failing for what seem to be unrelated reasons, so I'd like to focus on getting this working.

Kixunil commented:
Strangely when I wrote my test, I tried to check if it actually fails without the change and it didn't. I couldn't figure out why and then life dragged me into bunch of other things. I still intend to try resolve this but if someone can help that'd be appreciated.

Kixunil commented:
Figured it out - forgot to enable anchors :man_facepalming: 

I rebased the changes and added the test which fails before the fix and passes after the fix. I only need to add release notes, but I don't know which version. Would it be possible to add this to point release, please? I was originally hoping to get this into 0.14...

guggero commented:
Let's try with 0.14.1.

Kixunil commented:
Sorry for the noise, I had trouble setting up proper testing locally. Looks like those CI failures are unrelated, am I correct?

guggero commented:
No problem. Yes, unfortunately the CI pipeline is still a bit flaky. As long as most itest runs are green and the failures aren't related to any area you touched, they can be seen as flakes.

Kixunil commented:
Great, is there anything more that needs to be done?

guggero commented on lntest/itest/lnd_psbt_test.go:
```diff
@@ -470,6 +472,193 @@ func testPsbtChanFundingExternal(net *lntest.NetworkHarness, t *harnessTest) {
+	const chanSize = funding.MaxBtcFundingAmount
+
+	args := nodeArgsForCommitType(lnrpc.CommitmentType_ANCHORS)
+	// First, we'll create two new nodes that we'll use to open channels
```
nit: add newline before comment.

guggero commented on lnwallet/wallet.go:
```diff
@@ -706,9 +706,17 @@ func (l *LightningWallet) handleFundingReserveRequest(req *InitFundingReserveMsg
 		return
 	}
 
+	// We need to avoid enforcing reserved value in the middle of PSBT
```
nit: prefix commit message with affected package name.

guggero commented on lntest/itest/lnd_psbt_test.go:
```diff
@@ -470,6 +472,193 @@ func testPsbtChanFundingExternal(net *lntest.NetworkHarness, t *harnessTest) {
+	defer shutdownAndAssert(net, t, dave)
+
+	// Third node is only used for its wallet with PSBT funding abilities
+	funder := net.NewNode(t.t, "funder", nil)
```
Can we use `net.Alice` here instead? That node already has coins. This prevents us from creating too many new nodes (which always takes time and mines a bunch of blocks).

guggero commented on lntest/itest/lnd_psbt_test.go:
```diff
@@ -470,6 +472,193 @@ func testPsbtChanFundingExternal(net *lntest.NetworkHarness, t *harnessTest) {
+
+	// Before we start the test, we'll ensure both sides are connected so
+	// the funding flow can be properly executed.
+	_, cancel := context.WithTimeout(ctxb, defaultTimeout)
```
Can remove this line and the next one.

guggero commented on lntest/itest/lnd_psbt_test.go:
```diff
@@ -470,6 +472,193 @@ func testPsbtChanFundingExternal(net *lntest.NetworkHarness, t *harnessTest) {
+
+	net.SendCoins(t.t, btcutil.SatoshiPerBitcoin, funder)
+
+	// get new address for reserve
```
nit: comments should always be full sentences, including capitalization and punctuation. Previous comment and some further below as well.

guggero commented on lntest/itest/lnd_psbt_test.go:
```diff
@@ -470,6 +472,193 @@ func testPsbtChanFundingExternal(net *lntest.NetworkHarness, t *harnessTest) {
+		PkScript: reserveAddrScript,
+	}
+
+	decodedPsbt.UnsignedTx.TxOut = append(decodedPsbt.UnsignedTx.TxOut, &reserveTxOut)
```
nit: line length.

guggero commented on lntest/itest/lnd_psbt_test.go:
```diff
@@ -470,6 +472,193 @@ func testPsbtChanFundingExternal(net *lntest.NetworkHarness, t *harnessTest) {
+		Memo:  "new chans",
+		Value: int64(payAmt),
+	}
+	ctxt, _ = context.WithTimeout(ctxb, defaultTimeout)
```
nit: always add the `cancel` and `defer cancel()` for new code.

Kixunil replied:
FTR, I copied it from the test above where it's the same, so probably that will need fix too.

guggero commented on lntest/itest/lnd_psbt_test.go:
```diff
@@ -470,6 +472,193 @@ func testPsbtChanFundingExternal(net *lntest.NetworkHarness, t *harnessTest) {
+// wallet of both nodes are empty and one of them uses PSBT and an external
+// wallet to fund the channel while creating reserve output in the same
+// transaction.
+func testPsbtChanFundingSingleStep(net *lntest.NetworkHarness, t *harnessTest) {
```
Nice test!

guggero reviewed (commented):
Thank you for the fix! This will definitely make it easier to use the PSBT API.

Main comments are about some optimizations in the itest and the commit messages (prefixes mostly). See contribution guidelines for more information.

Kixunil commented:
Fixed hopefully everything. Sadly I got to even more trouble with my machine so I'll have to let the CI run the tests. Let's hope for the best.

guggero commented:
Ah, needs a rebase because of the release notes as well.

Kixunil commented:
LOL, was looking at test and thinking why they don't run yet. :D Done.

Kixunil commented:
Looks like same unrelated/flakes again.

Kixunil commented:
Updated release notes version to 0.14.2

guggero reviewed (approved):
LGTM :tada:

Kixunil commented:
Rebased

Roasbeef reviewed (approved):
LGTM 🎎

Kixunil commented:
Unrelated test failures again.
